### ISSUE TITLE
Recording files from Expo AV get corrupted

### ISSUE THREAD

**Issue by andordavoti** [expo/expo#25842]
### Minimal reproducible example

https://github.com/andordavoti/expo-av-issue-reproduction

### Issue

When a recording is stopped, expo-av saves it to the cache directory. I then move it to the document directory with expo-file-system so it's permanently there. However, after an app update through TestFlight or recompiling the app, it can't play the previously saved recordings.

Steps to reproduce in development:
1. Start the app with `yarn ios`
2. Create a couple recordings
3. Kill the app and dev server
4. Start the app again with `yarn ios`
5. The app can no longer play the recordings.

I've also tested this with production build through TestFlight. After an app update the recordings are unplayable.

### Logs from Xcode

I tried debugging this through Xcode and got the following:

> URL:file:///Users/andordavoti/Library/Developer/CoreSimulator/Devices/D5A3D6FD-F6D4-4765-8F08-5C5126C3933D/data/Containers/Data/Application/0251342E-79B2-4E55-B7A1-DC8DE91698AB/Documents/recording-C1116460-2D7B-43CD-A8D9-25A606D6C747.m4a : Error Domain=NSCocoaErrorDomain Code=257 "The file couldn’t be opened because you don’t have permission to view it."

It seems like I don't have permission to view the file. But it's stored in the app sandbox document directory, I'm not trying to access files outside what my app has access to, this should work...

### Logs from expo-av

Expo AV says the file isn't readable. Sounds like the file has corrupted for some reason. It also is impossible to export using react-native-share (after pressing "save to files" it just closes the file explorer before it can properly open it).

> File 'file:///Users/andordavoti/Library/Developer/CoreSimulator/Devices/D5A3D6FD-F6D4-4765-8F08-5C5126C3933D/data/Containers/Data/Application/46923CAB-0334-4C84-879E-2416C4A2F550/Documents/recording-DFF3B0D5-22F2-43A6-A1EC-8C3CCF7E7B9E.m4a' isn't readable.

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 14.1.2
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.17.0 - ~/.nvm/versions/node/v18.17.0/bin/node
      Yarn: 1.22.19 - ~/.yarn/bin/yarn
      npm: 9.6.7 - ~/.nvm/versions/node/v18.17.0/bin/npm
      Watchman: 2023.11.27.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.12.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
      Android SDK:
        API Levels: 28, 29, 30, 32, 33
        Build Tools: 29.0.2, 29.0.3, 30.0.2, 30.0.3, 31.0.0, 32.1.0, 33.0.0, 33.0.1, 33.0.2
        System Images: android-28 | ARM 64 v8a, android-32 | Google APIs ARM 64 v8a, android-32 | Google Play ARM 64 v8a, android-33 | Google Play ARM 64 v8a, android-34 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2023.1 AI-231.9392.1.2311.11076708
      Xcode: 15.0.1/15A507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~49.0.15 => 49.0.21 
      react: 18.2.0 => 18.2.0 
      react-native: 0.72.6 => 0.72.6 
    npmGlobalPackages:
      eas-cli: 5.9.1
    Expo Workflow: managed

**Comment by akodkod** [comment]
@andordavoti I'm getting the same issue. Did you find a solution?

**Comment by etticat** [comment]
We're running into the same issue

**Comment by uoon-dev** [comment]
same here

**Comment by uoon-dev** [comment]
I found another solution.
If you do the recording via expo-av and then call the stopAndUnloadAsync() function, the uri of the recorded data is set under /Library/Caches/AV and the audio file is created in that path. However, the existing /Library/Caches/AV path becomes inaccessible when the app is updated, so the recorded files that played well before the app update become unplayable after the update.

To solve this problem, I used the expo-file-system package to bypass the /Library/Caches/AV path by moving the recording files under the /Documents path. The code looks like this

```ts
import * as FileSystem from 'expo-file-system';
import { Audio } from 'expo-av';

const DOCUMENT_FOLDER = `${FileSystem.documentDirectory}`;

const saveRecording = async (uri: string) => {  // Recording.getURI()
  // Get just the name and extension of the recording file created from the URI path. eg) ephisa-wjfwanjdn.m4a 
  const fileName = uri.match(/\/Library\/Caches\/AV\/([^\/]+)$/)[1]; // 

  // Move the file that were in the old URI to the Documents folder.
  await FileSystem.moveAsync({ from: uri, to: `${DOCUMENT_FOLDER}${fileName}` }); // /Documents
}

const playRecording = async (fileName: string) => {  // Recording.getURI()
  const sound = new Audio.Sound();
  sound.loadAsync({ uri: `${DOCUMENT_FOLDER}${fileName}` });
  sound.playAsync();
}
```

By writing it like this, you can access the recording uri path under the Documents folder normally, even if you update the app. However, this method is used when the goal is to store recordings permanently in Documents rather than temporarily, as it does not store the recordings in the cache. Hope this helps.

**Comment by andordavoti** [comment]
@tsapeta any updates on this?

**Comment by vonovak** [comment]
Hello and thanks for reporting,
tl;dr: the solution to the issue is presented in the comment above.

To expand on this: you're seeing issues with file access (permissions, file not readable). The problem is that Xcode creates a new application ID every time you run the app. This is the expected behavior of Xcode and the solution is to not store _full absolute path_ because it becomes invalid.

However, in your repro the absolute url is created [here](https://github.com/andordavoti/expo-av-issue-reproduction/blob/73aed079ce727d42449d5164717cbd0c04b5228c/utils/recorderUtils.ts#L22) and persisted in asyncStorage [here](https://github.com/andordavoti/expo-av-issue-reproduction/blob/73aed079ce727d42449d5164717cbd0c04b5228c/components/Recorder.tsx#L62). The solution is to store inly file names and construct the complete uri as in the comment above:

```
const playRecording = async (fileName: string) => {  // eg. ephisa-wjfwanjdn.m4a 
  const sound = new Audio.Sound();
  await sound.loadAsync({ uri: `${FileSystem.documentDirectory}${fileName}` });
  await sound.playAsync();
}
```

I'm going to add a check to the `loadAsync` function to ensure we give an actionable warning in this case.

Hope this helps 🙂

**Comment by andordavoti** [comment]
Thank you guys so much! This helped me out a lot:) This would be great to have in the documentation for sure. I'll close this as it's working as intended.
